### ISSUE TITLE
PerlIO::via leaks memory at a high rate

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#9493]
Migrated from [rt.perl.org#59268](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=59268) (status was 'resolved')

Searchable as RT59268$

**Comment by p5pRT** [comment]
### From pcg@goof.com

#### Created by pcg@goof.com



The following test program leaks memory at a high rate \(every write\)&#8203;:

&nbsp;   while \(\) \{
&nbsp;      open my $fh\, ">&#8203;:via\(bla\)"\, \\my $buf or die;
&nbsp;      print $fh "x";
&nbsp;   \}  

&nbsp;   package PerlIO&#8203;::via&#8203;::bla;

&nbsp;   sub PUSHED \{
&nbsp;      \_\_PACAKGE\_\_
&nbsp;   \}

&nbsp;   sub WRITE \{
&nbsp;      \(print \{$\_\[2\]\} $\_\[1\]\) ? length $\_\[1\] : \-1
&nbsp;   \}  

An mtrace of a program running the loop some 400k times shows these leaks
as main causes\, but there are many smaller leaks in via\.so&#8203;:

&nbsp;   \[18065877 bytes in 403594 blocks\, 1st\. block is 45 bytes at 0x15c54a50\] 
&nbsp;      0x1da54cb2 libktrace\.so
&nbsp;      0x0045d085 util\.c&#8203;:95\(Perl\_safesysmalloc\)
&nbsp;      0x0046dc1d hv\.c&#8203;:2418\(S\_share\_hek\_flags\)
&nbsp;      0x0046f5da hv\.c&#8203;:766\(Perl\_hv\_common\)
&nbsp;      0x0046fe2c hv\.c&#8203;:323\(Perl\_hv\_common\_key\_len\)
&nbsp;      0x00541437 gv\.c&#8203;:1065\(Perl\_gv\_fetchpvn\_flags\)
&nbsp;      0x1fbe2c95 via\.so\(PerlIOVia\_method\+0x345\)
&nbsp;      0x1fbe30e5 via\.so\(PerlIOVia\_write\+0x65\)
&nbsp;      0x004e2fed doio\.c&#8203;:1247\(Perl\_do\_print\)
&nbsp;      0x00485f6f pp\_hot\.c&#8203;:778\(Perl\_pp\_print\)
&nbsp;      0x0047d27e run\.c&#8203;:38\(Perl\_runops\_standard\)
&nbsp;      0x0047975b perl\.c&#8203;:2384\(S\_run\_body\)
&nbsp;      0x0041f574 perlmain\.c&#8203;:113\(main\)
&nbsp;      0x1e6ac1a6 libc\.so\.6\(\_\_libc\_start\_main\+0xe6\)
&nbsp;   \[9726720 bytes in 2384 blocks\, 1st\. block is 4080 bytes at 0x03da8e70\] 
&nbsp;      0x1da54cb2 libktrace\.so
&nbsp;      0x0045d085 util\.c&#8203;:95\(Perl\_safesysmalloc\)
&nbsp;      0x00488718 sv\.c&#8203;:693\(Perl\_get\_arena\)
&nbsp;      0x0046f723 hv\.c&#8203;:44\(S\_more\_he\)
&nbsp;      0x0046fe2c hv\.c&#8203;:323\(Perl\_hv\_common\_key\_len\)
&nbsp;      0x00541437 gv\.c&#8203;:1065\(Perl\_gv\_fetchpvn\_flags\)
&nbsp;      0x1fbe2c95 via\.so\(PerlIOVia\_method\+0x345\)
&nbsp;      0x1fbe30e5 via\.so\(PerlIOVia\_write\+0x65\)
&nbsp;      0x004e2fed doio\.c&#8203;:1247\(Perl\_do\_print\)
&nbsp;      0x00485f6f pp\_hot\.c&#8203;:778\(Perl\_pp\_print\)
&nbsp;      0x0047d27e run\.c&#8203;:38\(Perl\_runops\_standard\)
&nbsp;      0x0047975b perl\.c&#8203;:2384\(S\_run\_body\)
&nbsp;      0x0041f574 perlmain\.c&#8203;:113\(main\)
&nbsp;      0x1e6ac1a6 libc\.so\.6\(\_\_libc\_start\_main\+0xe6\)

If I had to make a very quick guess\, could it be that the
pseudo\-filehandle creates via newGVgen creates a new \_GEN\_xxx symbol in
the \.\.\.&#8203;::bla package that never ever gets freed?

Indeed\, adding this line into the while loop&#8203;:

&nbsp;   warn keys %PerlIO&#8203;::via&#8203;::bla&#8203;::;

creates successively longer output lines\. Could that mean that
newGVgen leaks?

<details><summary>Perl Info</summary>

```

Flags:
    category=library
    severity=medium

Site configuration information for perl 5.10.0:

Configured by Marc Lehmann at Sat Jan 19 07:46:38 CET 2008.

Summary of my perl5 (revision 5 version 10 subversion 0 patch 33007) configuration:
  Platform:
    osname=linux, osvers=2.6.23-1-amd64, archname=amd64-linux
    uname='linux cerebro 2.6.23-1-amd64 #1 smp fri dec 21 12:00:17 utc 2007 x86_64 gnulinux '
    config_args='-Duselargefiles -Dxxxxuse64bitint -Uuse64bitall -Dusemymalloc=n -Dcc=gcc -Dccflags=-DPERL_DONT_CREATE_GVSV_disabled -ggdb -Dcppflags=-DPERL_DONT_CREATE_GVSV_disabled -D_GNU_SOURCE -I/opt/include -Doptimize=-O6 -march=opteron -mtune=opteron -funroll-loops -fno-strict-aliasing -Dcccdlflags=-fPIC -Dldflags=-L/opt/perl/lib -L/opt/lib -Dlibs=-ldl -lm -lcrypt -Darchname=amd64-linux -Dprefix=/opt/perl -Dprivlib=/opt/perl/lib/perl5 -Darchlib=/opt/perl/lib/perl5 -Dvendorprefix=/opt/perl -Dvendorlib=/opt/perl/lib/perl5 -Dvendorarch=/opt/perl/lib/perl5 -Dsiteprefix=/opt/perl -Dsitelib=/opt/perl/lib/perl5 -Dsitearch=/opt/perl/lib/perl5 -Dsitebin=/opt/perl/bin -Dman1dir=/opt/perl/man/man1 -Dman3dir=/opt/perl/man/man3 -Dsiteman1dir=/opt/perl/man/man1 -Dsiteman3dir=/opt/perl/man/man3 -Dman1ext=1 -Dman3ext=3 -Dpager=/usr/bin/less -Uafs -Uusesfio -Uusenm -Uuseshrplib -Dd_dosuid -Dusethreads=undef -Duse5005threads=undef -Duseithreads=undef -Dusemultiplicity=undef -Demail=perl-binary@plan9.de -Dcf_email=perl-binary@plan9.de -Dcf_by=Marc Lehmann -Dlocincpth=/opt/perl/include /opt/include -Dmyhostname=localhost -Dmultiarch=undef -Dbin=/opt/perl/bin -Dxxxusedevel -des'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags ='-DPERL_DONT_CREATE_GVSV_disabled -ggdb -fno-strict-aliasing -pipe -I/opt/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O6 -march=opteron -mtune=opteron -funroll-loops -fno-strict-aliasing',
    cppflags='-DPERL_DONT_CREATE_GVSV_disabled -D_GNU_SOURCE -I/opt/include -DPERL_DONT_CREATE_GVSV_disabled -ggdb -fno-strict-aliasing -pipe -I/opt/include'
    ccversion='', gccversion='4.2.3 20080102 (prerelease) (Debian 4.2.2-5)', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags ='-L/opt/perl/lib -L/opt/lib -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib /lib64 /usr/lib64
    libs=-ldl -lm -lcrypt
    perllibs=-ldl -lm -lcrypt
    libc=/lib/libc-2.7.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.7'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O6 -march=opteron -mtune=opteron -funroll-loops -fno-strict-aliasing -L/opt/perl/lib -L/opt/lib -L/usr/local/lib'

Locally applied patches:
    MAINT


@INC for perl 5.10.0:
    /root/src/sex
    /opt/perl/lib/perl5
    /opt/perl/lib/perl5
    /opt/perl/lib/perl5
    /opt/perl/lib/perl5
    /opt/perl/lib/perl5
    .


Environment for perl 5.10.0:
    HOME=/root
    LANG (unset)
    LANGUAGE (unset)
    LC_CTYPE=de_DE.UTF-8
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/root/s2:/root/s:/opt/bin:/opt/sbin:/bin:/sbin:/usr/bin:/usr/sbin:/usr/X11/bin:/usr/games:/usr/local/bin:/usr/local/sbin:/root/pserv:.
    PERL5LIB=/root/src/sex
    PERL5_CPANPLUS_CONFIG=/root/.cpanplus/config
    PERLDB_OPTS=ornaments=0
    PERL_ANYEVENT_EDNS0=1
    PERL_ANYEVENT_NET_TESTS=1
    PERL_ANYEVENT_PROTOCOLS=ipv4
    PERL_ANYEVENT_STRICT=1
    PERL_BADLANG (unset)
    PERL_UNICODE=0
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From chrmue@cpan.org

Yes\, this memory leak exists because of the stash entries generated by 
newGVgen\(\)\. After destroying the tied handle\, the symbol is not been 
removed from the stash hv\. If you try removing the entry from the stash 
by yourself you will get a warning "unable to free unreferenced 
scalar"\. Dear Perl\, what's the reason for this problem? How can we 
solve it?

**Comment by p5pRT** [comment]
### From chrmue@cpan.org

Sorry for the mistake\. I got this problem with tied handles\. It appears 
with every call of newGVgen\(\)\. The fact is\, it's not possible to 
generate tied handles \(in XS\) without generating memory leaks too\.

**Comment by p5pRT** [comment]
### From @nwc10

On Wed\, Sep 24\, 2008 at 01&#8203;:40&#8203;:20AM \-0700\, Marc Lehmann wrote&#8203;:

Sorry about the delay in anyone replying\. I don't think that anyone \(alive\)
is familiar with this code\.

Having been forced to have a look at it \(5\.8\.9 show stopper\) I can't say that
it fills me with joy\.

> If I had to make a very quick guess\, could it be that the
> pseudo\-filehandle creates via newGVgen creates a new \_GEN\_xxx symbol in
> the \.\.\.&#8203;::bla package that never ever gets freed?
> 
> Indeed\, adding this line into the while loop&#8203;:
> 
>    warn keys %PerlIO&#8203;::via&#8203;::bla&#8203;::;
> 
> creates successively longer output lines\. Could that mean that
> newGVgen leaks?

I am suspicious of this part of the code of via\.xs

&nbsp;        if \(\!s\->fh\) \{
&nbsp;        GV \*gv = newGVgen\(HvNAME\_get\(s\->stash\)\);
&nbsp;        GvIOp\(gv\) = newIO\(\);
&nbsp;        s\->fh = newRV\(\(SV \*\) gv\);
&nbsp;        s\->io = GvIOp\(gv\);
&nbsp;        \}


I suspect that yes\, it practical terms it is leaking at the Perl level\, in
that all the memory is still reachable\, and will be correctly accounted for
at interpreter exit\, but the code never thinks to clean up its temporaries
immediately after it has finished with them\.

I'm afraid that I'm not going to have time to make a fix that I'm comfortable
with in time for 5\.8\.9\. However\, I don't see any particular reason why a
fixed version can't be put on CPAN subsequently\, which would allow you \(or
anyone else\) to have a dependency on it\, and hence "fix" existing perl
installations\.

Nicholas Clark

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @mhx

On 2008\-10\-26\, at 17&#8203;:41&#8203;:26 \+0000\, Nicholas Clark wrote&#8203;:

> On Wed\, Sep 24\, 2008 at 01&#8203;:40&#8203;:20AM \-0700\, Marc Lehmann wrote&#8203;:
> 
> Sorry about the delay in anyone replying\. I don't think that anyone \(alive\)
> is familiar with this code\.
> 
> Having been forced to have a look at it \(5\.8\.9 show stopper\) I can't say that
> it fills me with joy\.
> 
> > If I had to make a very quick guess\, could it be that the
> > pseudo\-filehandle creates via newGVgen creates a new \_GEN\_xxx symbol in
> > the \.\.\.&#8203;::bla package that never ever gets freed?
> > 
> > Indeed\, adding this line into the while loop&#8203;:
> > 
> >    warn keys %PerlIO&#8203;::via&#8203;::bla&#8203;::;
> > 
> > creates successively longer output lines\. Could that mean that
> > newGVgen leaks?
> 
> I am suspicious of this part of the code of via\.xs
> 
>         if \(\!s\->fh\) \{
>         GV \*gv = newGVgen\(HvNAME\_get\(s\->stash\)\);
>         GvIOp\(gv\) = newIO\(\);
>         s\->fh = newRV\(\(SV \*\) gv\);
>         s\->io = GvIOp\(gv\);
>         \}

The trick is to get rid of the GV reference created in the
stash immediately\, as it will otherwise survive the death
of the layer\. This is exactly what IO&#8203;::File does in its
implementation of new\_tmpfile\(\)\.

> I suspect that yes\, it practical terms it is leaking at the Perl level\, in
> that all the memory is still reachable\, and will be correctly accounted for
> at interpreter exit\, but the code never thinks to clean up its temporaries
> immediately after it has finished with them\.
> 
> I'm afraid that I'm not going to have time to make a fix that I'm comfortable
> with in time for 5\.8\.9\. However\, I don't see any particular reason why a
> fixed version can't be put on CPAN subsequently\, which would allow you \(or
> anyone else\) to have a dependency on it\, and hence "fix" existing perl
> installations\.

Agreed\. I've made the change below in blead so we can see
if it causes any problems\. If it turns out to be fine\, a
PerlIO&#8203;::via release should fix it for older perls\.

What remains is the question whether we can avoid adding
the GV to the stash in the first place\. Maybe by passing
yet another flag to Perl\_gv\_fetchpvn\_flags\(\)?

Marcus

PS&#8203;: Yes\, the "if \(gv\)" is dumb\. It'll never get there if
&nbsp;    gv == NULL\. Just noticed after committing\.\.\.


Change 34779 by mhx@&#8203;mhx\-r2d2 on 2008/11/08 18&#8203;:31&#8203;:09

&nbsp;    Make sure PerlIO&#8203;::via doesn't leave its old GV's hanging
&nbsp;    in the stash they were created in\. Steal the idea from
&nbsp;    IO&#8203;::File to just delete the stash entry after creating
&nbsp;    the GV\.
&nbsp;    This fixes perl \#59268\.

Affected files \.\.\.

\.\.\. //depot/perl/ext/PerlIO/via/via\.pm\#11 edit
\.\.\. //depot/perl/ext/PerlIO/via/via\.xs\#18 edit

Differences \.\.\.

==== //depot/perl/ext/PerlIO/via/via\.pm\#11 \(text\) ====

@&#8203;@&#8203; \-1\,5 \+1\,5 @&#8203;@&#8203;
 package PerlIO&#8203;::via;
\-our $VERSION = '0\.06';
\+our $VERSION = '0\.07';
 use XSLoader \(\);
 XSLoader&#8203;::load 'PerlIO&#8203;::via';
 1;

==== //depot/perl/ext/PerlIO/via/via\.xs\#18 \(text\) ====

@&#8203;@&#8203; \-91\,6 \+91\,10 @&#8203;@&#8203;
&nbsp;         GvIOp\(gv\) = newIO\(\);
&nbsp;         s\->fh = newRV\(\(SV \*\) gv\);
&nbsp;         s\->io = GvIOp\(gv\);
\+        if \(gv\) \{
\+            /\* shamelessly stolen from IO&#8203;::File's new\_tmpfile\(\) \*/
\+            hv\_delete\(GvSTASH\(gv\)\, GvNAME\(gv\)\, GvNAMELEN\(gv\)\, G\_DISCARD\);
\+        \}
&nbsp;         \}
&nbsp;         IoIFP\(s\->io\) = PerlIONext\(f\);
&nbsp;         IoOFP\(s\->io\) = PerlIONext\(f\);

**Comment by p5pRT** [comment]
### From @mhx

Fixed by change \#34779\.

**Comment by p5pRT** [comment]
@mhx - Status changed from 'open' to 'resolved'
